### ISSUE TITLE
ActivitySummary's receivers.py and User.delete()

### ISSUE THREAD

**Issue by jup014** [kpwhri/heartsteps#226]
tl;dr - Because of **day_summary_updates_activity_summary()** receiver, ActivitySummary model’s user CASCADE option does not work properly.

I am not sure if this is a bug, but I think is notable enough to report.

See also: https://github.com/kpwhri/heartsteps/blob/master/server/activity_summaries/receivers.py#L60

Here's my guess about how the bug is produced.

1. If we delete a User who has an ActivitySummary, the instance of the User model is firstly gone.
2. Because the ActivitySummary instance is referring to the user instance with FK under CASCADE criteria, it is deleted too.
3. But then, an instance of ActivitySummary is deleted, the receiver ( day_summary_updates_activity_summary() ) is triggered.
4. In the receiver, it automatically creates another ActivitySummary instance using the pre-stored **user_id** value in another deleted **Day** instance. You know, even if an instance is deleted from the database, the _xxxx_id_ of other referring instances is not invalidated (it's just an integer). Thus, a new instance of ActivitySummary is successfully created again with the already-deleted User's id, which causes IntegrityError.

I'm not sure what exactly this code purposes, but it definitely limits the testing of ActivitySummary class. 

There are two possible possibilities: (a) I'm not deleting instances in the right order (b) I'm supposed to do something before I delete the User instance.

An easy way to solve this is we can check if the cause of the trigger of the receiver is the deletion of the parent User instance. If so, we can skip the creation of the new instance of ActivitySummary. Otherwise, we can create as we do now. 

But because I don't understand the true purpose of this code, I'm not 100% sure if this is a good workaround to avoid this issue. Can anybody help?

**Comment by jup014** [comment]
I looked into it and I found that I was confused a little bit. Here is the correction:

---

tl;dr - Because of **day_summary_updates_activity_summary()** receiver, ActivitySummary model’s user CASCADE option does not work properly.

I am not sure if this is a bug, but I think is notable enough to report.

See also: https://github.com/kpwhri/heartsteps/blob/master/server/activity_summaries/receivers.py#L60

Here's my guess about how the bug is produced.

1. If we delete a User who has a **Day** instance(s), the instance of the User model is firstly gone.
2. If there were **ActivitySummary** instances referring to the user, they will be deleted due to the CASCADE FK. If not, it doesn't matter. What matters is, there will be no ActivitySummary after the cascade deletions.
3. If there were **Day** instances referring to the user, they will be deleted due to the CASCADE FK. This deletion triggers the receiver ( day_summary_updates_activity_summary() ).
4. In the receiver, it automatically creates another ActivitySummary instance using the **pre-stored user_id value** in **the deleted Day instance**. You know, even if an instance is deleted from the database, the _xxxx_id's_ in other referring instances are not invalidated (they're just integers). Thus, a new instance of ActivitySummary is _successfully_ created again with the already-deleted User's id, which causes IntegrityError.

I'm not sure what exactly this code purposes, but it definitely limits the testing of ActivitySummary and Day class; we need to delete the test user in tearDown().

There are two possibilities: (a) I'm not deleting instances in the right order (b) I'm supposed to do something before I delete the User instance.

An easy way to solve this is we can check if the cause of the trigger of the receiver is the deletion of the parent User instance. If so, we can skip the creation of the new instance of ActivitySummary. Otherwise, we can create as we do now.

But because I don't understand the true purpose of this code, I'm not 100% sure if this is a good workaround to avoid this issue. Can anybody help?

**Comment by nickdotreid** [comment]
This is a good discussion -- I don't 100% follow your reproduction steps.

An existing bug that I haven't written down, is that on the KPW system it's not possible to delete a participant or user account. I'm unsure if that is related to the issue you are discussing here.

Let's make a point to discuss this after you are back from vacation.
